File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int n, int[][] edges) {
        List<List<Integer>> adj = new ArrayList<>(n);
        int[] indeg = new int[n];
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        Arrays.fill(indeg , 0);
        for (var e : edges) {
            adj.get(e[1]).add(e[0]);
            indeg[e[0]]++;
        } 
        Deque<Integer> q = new ArrayDeque<>();
        for (int u=0 ; u<n ; u++) {
            if (indeg[u] == 0) q.addLast(u);
        }
        List<Integer> ans = new ArrayList<>();
        while (!q.isEmpty()) {
            int u = q.removeFirst();
            ans.add(u);
            for (int v : adj.get(u)) {
                indeg[v]--;
                if (indeg[v] == 0) q.addLast(v);
            }
        }
        if (ans.size() != n) return new int[0];
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            res[i] = ans.get(i);
        }
        return res;
    }
}

File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int n, vector<vector<int>>& edges) {
        vector<int> adj[n];
        vector<int> indeg(n , 0);
        for (auto e : edges) {
            adj[e[1]].push_back(e[0]);
            indeg[e[0]]++;
        }
        // for (auto i : indeg) cout << i << ' ';
        // cout << '\n';
        vector<int> ans;
        queue<int> q;
        for (int u=0 ; u<n ; u++) {
            if (indeg[u]) continue;
            q.push(u);
        }
        while (!q.empty()) {
            int u = q.front(); q.pop();
            ans.push_back(u); 
            for (int v : adj[u]) {
                indeg[v]--;
                if (!indeg[v]) q.push(v);
            }
        }
        if (ans.size() != n) return {};
        return ans;
    }
};